### PR TITLE
Remove `CircSupplyCalc` Trait

### LINKED ISSUE #2031: Remove the `CircSupplyCalc` trait.
**Issue summary**

The `CircSupplyCalc` no longer serves a purpose and should be removed. In `VM::new`, it can be replaced by a `FnOnce` callback that computes the circulating supply.

**Other information and links**

https://github.com/ChainSafe/forest/blob/446bec41daa29d55ddfd3befa5e697d492302b89/vm/interpreter/src/vm.rs#L40-L47

### PR DESCRIPTION
**Pull request by jdjaustin**
Co-authored-by: David Himmelstrup <lemmih@gmail.com>

**Summary of changes**
Changes introduced in this pull request:
- The `CircSupplyCalc` no longer serves a purpose. In `VM::new`, replaced `CircSupplyCalc` by an `FnOnce` callback that computes the circulating supply.

**Reference issue to close (if applicable)**

Closes #2031 

**Other information and links**

N/A

### CODE REVIEW

File: blockchain/state_manager/src/vm_circ_supply.rs [hunk 1]
```diff
@@ -63,6 +62,15 @@ impl GenesisInfo {
...
         }
     }
+
+    // Allows generation of the current circulating supply
+    pub fn get_supply<DB: BlockStore>(
+        &self,
+        height: ChainEpoch,
+        state_tree: &StateTree<DB>,
+    ) -> Result<TokenAmount, anyhow::Error> {
+        get_circulating_supply(self, height, state_tree)
```
Comment (lemmih) [inline blockchain/state_manager/src/vm_circ_supply.rs:72]: `get_circulating_supply` isn't used anywhere else so we can just inline it here. And you can change the `get_supply` name to the slightly more descriptive `get_circulating_supply`.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: vm/interpreter/src/vm.rs [hunk 2]
```diff
@@ -98,14 +87,14 @@ where
...
-    pub fn new<R, C>(
+    pub fn new<R>(
         root: Cid,
         store_arc: DB,
         epoch: ChainEpoch,
         rand: &R,
         base_fee: BigInt,
         network_version: NetworkVersion,
-        circ_supply_calc: C,
+        circ_supply_calc: impl FnOnce(ChainEpoch, &StateTree<&DB>) -> Result<TokenAmount, anyhow::Error>,
```
Comment (LesnyRumcajs) [inline vm/interpreter/src/vm.rs:97]: Let's not use the shorthand with such a long type. It'd be nicer to put it in the `where` section.

File: blockchain/state_manager/src/lib.rs [hunk 3]
```diff
@@ -1384,7 +1382,7 @@ where
         height: ChainEpoch,
         state_tree: &StateTree<&DB>,
     ) -> Result<TokenAmount, anyhow::Error> {
-        self.genesis_info.get_supply(height, state_tree)
+        self.genesis_info.get_circulating_supply(height, state_tree)
```
Comment (lemmih) [inline blockchain/state_manager/src/lib.rs:1385]: I would have liked this to be inlined such that there is only one `get_circulating_supply` function. But this is also okay.

**Review by lemmih — APPROVED** [review]

**Review by LesnyRumcajs — APPROVED** [review]
I like it.

**Pull request merged**